packages/fe-fpm-writer/src/controller-extension/index.ts

```diff
@@ -0,0 +1,207 @@
+        ...data
+    };
+    // Apply default data
+    setCommonDefaults(config, manifestPath, manifest);
+    // Create `controllerName` with full path/namespace
+    config.controllerName = `${config.ns}.${config.name}`;
+    // Resolve controller extension id/key
+    let extensionId = EXTENSION_PAGE_TYPE_MAP.get(config.pageType) || UI5_CONTROLLER_EXTENSION_LIST_REPORT;
+    if (config.pageId) {
+        // Prepend project id
+        extensionId = `${extensionId}#${manifest['sap.app'].id}::${config.pageId}`;
+    }
+    config.extensionId = extensionId;
+    // Get existing controller extension entry from manifest
+    const manifestExtension = getExistingControllerExtension(manifest, extensionId);
+    // If controller extension already exists in manifest - append new controller
+    if (manifestExtension) {
+        handleExistingManifestExtension(
+            manifestExtension,
+            config as InternalControllerExtension,
+            config.controllerName
+        );
+    }
+
+    return config as InternalControllerExtension;
+}
+
+/**
+ * A function that transforms JSON object during JSON.stringify call.
+ * Method is used to remove 'controllerName' or 'controllerNames' properties from manifest in case when we have transition from 'controllerName' to 'controllerNames'.
+ *
+ * @param {InternalControllerExtension} config - a controller extension configuration object
+ * @returns Json replacer method
+ */
+function getManifestReplacer(
+    config: InternalControllerExtension
+): ((key: string, value: Manifest) => unknown) | undefined {
+    let isRoot = true;
+    const { deleteProperty } = config;
+    if (!deleteProperty) {
+        // No request to delete any property
+        return undefined;
+    }
+    return (key: string, value: Manifest) => {
+        // Handle only root - more stable solution instead of checking 'key'
+        if (key === '' && isRoot) {
+            isRoot = false;
+            const extension = getExistingControllerExtension(value, config.extensionId);
+            if (extension) {
+                delete extension[deleteProperty];
+            }
+        }
+        return value;
+    };
+}
+
+/**
+ * Add a controller extension to an existing UI5 application.
+ *
+ * @param {string} basePath - the base path
+ * @param {ControllerExtension} controllerConfig - the controller extension configuration
+ * @param {Editor} [fs] - the memfs editor instance
+ * @returns {Editor} the updated memfs editor instance
+ */
+export function generateControllerExtension(
+    basePath: string,
+    controllerConfig: ControllerExtension,
+    fs?: Editor
+): Editor {
+    // Validate the base and view paths
+    if (!fs) {
+        fs = create(createStorage());
+    }
+    validateBasePath(basePath, fs);
+
+    const manifestPath = join(basePath, 'webapp/manifest.json');
+    const manifest = fs.readJSON(manifestPath) as Manifest;
+
+    const root = join(__dirname, '../../templates');
```

Comment:
With https://github.com/SAP/open-ux-tools/commit/a1dc0698b66fa8fbfcf7523488c0b5f1afc440f8 we have a common function to access the template folders (makes it easier to integrate it when webpacking). Please merge from main and use it instead of determining root here.

Reply:
Thanks for [6071127](https://github.com/SAP/open-ux-tools/pull/653/commits/6071127d76c268b0938e66a3550eaa2b63586172)